### DISCUSSION TITLE
Proper way to modify player inventory contents

### DISCUSSION THREAD

**Discussion by KianBrose** [Minestom/Minestom#1799] [Help] [answered]
Context: I am migrating a gun plugin from PaperMC to Minestom and have now reached the point where I am giving the guns an ammunition management system. Each item has a specific type of bullet, such as `Material.FLINT` and a magazine size like for example `30`. What I am trying to achieve is a function capable of 
1. Reading the player inventory and filtering the items of that bullet type
2. Removing a quantity of those items up to 30

The problem I am facing is when it comes to removing the items. I am aware that ItemStacks are immutable and cannot be changed, what I did after some testing is as follows:
```java
Material bulletMaterial = Material.fromNamespaceId(bulletType);
Stream<ItemStack> itemStackStream = Arrays.stream(inventory.getItemStacks()).filter(ev -> ev.material() == bulletMaterial);
var itemIter = itemStackStream.iterator();
while (itemIter.hasNext()){
      ItemStack itemStac = itemIter.next();
      // I don't know here
}
```
This function is properly able to find the ItemStacks and quantities of said itemstacks in the player inventory, but where do I go from here to "update" that specific itemStack's quantity to a specific quantity X or remove it completely? I do not see an option of `itemStac` that lets me get the slot ID of that specific item which would give the the option to .setItem, so at the moment I'm a bit clueless

**Comment by GoldenStack** [comment]
Because items are immutable, data on them shouldn't tie them to any specific slot, as the same object could be in multiple slots.
You need to preserve the slot IDs when doing this operation.

**Comment by ZakShearman** [comment, marked as answer]
Not 100% on this but IIRC Minestom sets empty slots to an ItemStack of AIR. Therefore, you should be able to use the index of the ItemStack[] returned by getItemStacks on the inventory.

Therefore, you can iterate over this with a for loop with an index.

Try this out and let me know if it works :)

P.S just for future help, this is specifically what I mean as a demo in the form of a command:

```java
public class TestCommand extends Command {
    public TestCommand() {
        super("test");

        var materialArg = new ArgumentItemStack("material");
        var amountArg = new ArgumentInteger("amount").min(1);
        this.addSyntax(this::execute, materialArg, amountArg);
    }

    private void execute(CommandSender sender, CommandContext context) {
        ItemStack item = context.get("material");
        int amount = context.get("amount");

        if (!(sender instanceof Player player)) {
            sender.sendMessage("You must be a player to use this command");
            return;
        }

        this.takeAmmo(player, amount, item.material());
    }

    private void takeAmmo(Player player, int amount, Material material) {
        ItemStack[] items = player.getInventory().getItemStacks();
        for (int i = 0; i < items.length; i++) {
            ItemStack item = items[i];
            if (item.material() != material) continue;

            int stackSize = item.amount();
            if (stackSize < amount) {
                amount -= stackSize;
                player.getInventory().setItemStack(i, ItemStack.AIR);
            } else {
                player.getInventory().setItemStack(i, item.withAmount(stackSize - amount));
                break;
            }
        }
    }
}
```

**Comment by DReinholdtsen** [comment]
ZakShearman already answered your question quite well, but I will add some additional information. Since ItemStacks are immutable, and do not contain any slot data, you can't just loop over the ItemStacks in a player's inventory. You must additionally keep track of the current slot the ItemStack you are looking for is in. Once you have the slot of the item you want to modify, as well as the amount you want to modify it by (let's say one for this instance), you can use the code
`inventory.setItemStack(slot, originalItem.withAmount(originalItem.amount()-1))`
There are other ways of doing that, and you may want to get the original item amount and store it in a variable if you are going to use it again.
Alternatively, in this specific use case, you could have a separate inventory for each ammo type (this is just an implementation suggestion, and if you truly want to port the plugin over 1 to 1, then this isn't applicable)
This would make it much easier to keep track and take ammo from players.

**Comment by KianBrose** [comment]
Thank you ZakShearman and DReinholdtsen! The solutions worked wonderfully!
